File: app/src/main/java/com/reteno/sample/fragments/FragmentDeviceId.java
package com.reteno.sample.fragments;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.reteno.core.RetenoImpl;
import com.reteno.core._interop.DeviceIdInternal;
import com.reteno.core.data.local.config.DeviceIdMode;
import com.reteno.core.data.repository.ConfigRepository;
import com.reteno.core.di.ServiceLocator;
import com.reteno.sample.BaseFragment;
import com.reteno.sample.databinding.FragmentDeviceIdBinding;
import com.reteno.sample.util.AppSharedPreferencesManager;

import java.lang.reflect.Field;

import kotlin.Unit;

public class FragmentDeviceId extends BaseFragment {

    private FragmentDeviceIdBinding binding;
    private ServiceLocator serviceLocator;
    private ConfigRepository configRepository;

    public FragmentDeviceId() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            Field field = RetenoImpl.class.getDeclaredField("serviceLocator");
            field.setAccessible(true);
            serviceLocator = (ServiceLocator) field.get(getReteno());
            field.setAccessible(false);
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        configRepository = serviceLocator.getConfigRepositoryProvider().get();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentDeviceIdBinding.inflate(getLayoutInflater(), container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        initExternalDeviceId(view);
        refreshUi();

        initListeners(view);
    }

    private void initExternalDeviceId(@NonNull View view) {
        String externalSavedId = AppSharedPreferencesManager.getExternalId(view.getContext());
        if (!TextUtils.isEmpty(externalSavedId)) {
            getReteno().setUserAttributes(externalSavedId);
        }
    }

    private void refreshUi() {
        String id = DeviceIdInternal.INSTANCE.getIdInternal(configRepository.getDeviceId());
        String externalId = DeviceIdInternal.INSTANCE.getExternalIdInternal(configRepository.getDeviceId());
        DeviceIdMode mode = DeviceIdInternal.INSTANCE.getModeInternal(configRepository.getDeviceId());
        binding.tvCurrentDeviceIdMode.setText(mode.toString());
        binding.tvCurrentDeviceId.setText(id);
        binding.tvExternalId.setText(externalId);
        configRepository.getFcmToken(str -> {
                    binding.etFcmToken.setText(str);
                    return Unit.INSTANCE;
                }
        );
    }

    private void initListeners(@NonNull View view) {
        binding.rootView.setOnRefreshListener(() -> {
            refreshUi();
            binding.rootView.setRefreshing(false);
        });
        binding.tilExternalId.setStartIconOnClickListener(v -> {
            String externalId = binding.etExternalId.getText().toString();
            AppSharedPreferencesManager.saveExternalId(view.getContext(), externalId);
            getReteno().setUserAttributes(externalId);
            refreshUi();
        });
        binding.tilExternalId.setEndIconOnClickListener(v -> {
            AppSharedPreferencesManager.saveExternalId(view.getContext(), "");
            getReteno().setUserAttributes("");
            binding.etExternalId.setText("");
            refreshUi();
        });
    }
}

File: app/src/main/java/com/reteno/sample/fragments/FragmentDeviceId.kt
package com.reteno.sample.fragments

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.lifecycleScope
import com.reteno.core.RetenoInternalImpl
import com.reteno.core._interop.DeviceIdInternal.getExternalIdInternal
import com.reteno.core._interop.DeviceIdInternal.getIdInternal
import com.reteno.core._interop.DeviceIdInternal.getModeInternal
import com.reteno.core.data.repository.ConfigRepository
import com.reteno.core.di.ServiceLocator
import com.reteno.sample.BaseFragment
import com.reteno.sample.databinding.FragmentDeviceIdBinding
import com.reteno.sample.util.AppSharedPreferencesManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class FragmentDeviceId : BaseFragment() {

    private var binding: FragmentDeviceIdBinding? = null
    private var serviceLocator: ServiceLocator? = null
    private var configRepository: ConfigRepository? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            val field = RetenoInternalImpl::class.java.getDeclaredField("serviceLocator")
            field.isAccessible = true
            serviceLocator = field[reteno] as ServiceLocator
            field.isAccessible = false
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        configRepository = serviceLocator!!.configRepositoryProvider.get()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentDeviceIdBinding.inflate(layoutInflater, container, false)
        return binding!!.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initExternalDeviceId(view)
        refreshUi()
        initListeners(view)
    }

    private fun initExternalDeviceId(view: View) {
        val externalSavedId = AppSharedPreferencesManager.getExternalId(view.context)
        if (!TextUtils.isEmpty(externalSavedId)) {
            reteno.setUserAttributes(externalSavedId)
        }
    }

    private fun refreshUi() {
        val id = configRepository!!.getDeviceId().getIdInternal()
        val externalId = configRepository!!.getDeviceId().getExternalIdInternal()
        val mode = configRepository!!.getDeviceId().getModeInternal()
        binding!!.tvCurrentDeviceIdMode.text = mode.toString()
        binding!!.tvCurrentDeviceId.text = id
        binding!!.etDeviceId.setText(AppSharedPreferencesManager.getDeviceId(requireContext()))
        binding!!.etDeviceIdMillis.setText(
            AppSharedPreferencesManager.getDeviceIdDelay(
                requireContext()
            ).toString()
        )
        binding!!.tvExternalId.text = externalId
        lifecycleScope.launch {
            val token = withContext(Dispatchers.IO) {
                configRepository!!.getFcmToken()
            }
            binding!!.etFcmToken.setText(token)
        }
    }

    private fun initListeners(view: View) {
        binding!!.rootView.setOnRefreshListener {
            refreshUi()
            binding!!.rootView.isRefreshing = false
        }
        binding!!.tilExternalId.setStartIconOnClickListener {
            val externalId = binding!!.etExternalId.getText().toString()
            AppSharedPreferencesManager.saveExternalId(view.context, externalId)
            reteno.setUserAttributes(externalId)
            refreshUi()
        }
        binding!!.tilExternalId.setEndIconOnClickListener {
            AppSharedPreferencesManager.saveExternalId(view.context, "")
            reteno.setUserAttributes("")
            binding!!.etExternalId.setText("")
            refreshUi()
        }
        binding!!.tilDeviceId.setStartIconOnClickListener {
            val deviceId = binding!!.etDeviceId.getText().toString()
            AppSharedPreferencesManager.saveDeviceId(view.context, deviceId)
            refreshUi()
        }
        binding!!.tilDeviceId.setEndIconOnClickListener {
            AppSharedPreferencesManager.saveDeviceId(view.context, "")
            binding!!.etDeviceId.setText("")
            refreshUi()
        }
        binding!!.tilDeviceIdFetchDelay.setStartIconOnClickListener {
            val deviceIdMillis = binding!!.etDeviceIdMillis.getText().toString()
            AppSharedPreferencesManager.saveDeviceIdDelay(view.context, deviceIdMillis.toInt())
            refreshUi()
        }
        binding!!.tilDeviceIdFetchDelay.setEndIconOnClickListener {
            AppSharedPreferencesManager.saveDeviceIdDelay(view.context, 0)
            binding!!.etDeviceId.setText("")
            refreshUi()
        }
    }
}